Beginning removal of pycbc_glue

spxiwh opened the pull request:
This patch begins the second stage of the pycbc_glue -> glue transition. Here I:

 * Add lscsoft_glue as a dependancy (now on pypi)
 * Move all imports to pycbc_glue (except pycbc_glue.ligolw) to glue

I also have *removed* the mvsc executables as these use code (the SQL stuff in glue), that we have no intention currently to support within pycbc. (@tdent) Of course one can always access these through git history and/or create a small repository to host those if needed ... Or argue against my decision here.

In parallel I will open a patch in pycbc_glue to remove all of the non ligolw codes and then make a new pycbc_glue release.

spxiwh commented:
Fixes #2015

spxiwh commented:
Please hold up on merging this until we have a functional 1.9.X release.

spxiwh commented:
It looks like the Travis issues are not resolvable with the current version of `lscsoft-glue`. We'll have to put this on hold while working on an updated lscsoft-glue release.

spxiwh commented:
@ahnitz @duncanmmacleod This patch has now passed the Travis tests, so I'm removing the "On hold" tab and ready to ask if this can be merged

ahnitz commented:
@spxiwh, we need a new release of glue to not break pycbc workflows unfortunately.

spxiwh commented:
Could you explain here what the problem is @ahnitz ?

ahnitz commented:
@spxiwh So nominally, I approve this with the caveat it is tested on a test workflow and a new version of glue is release (and you update the dependency) to require it.Please merge once that's satisfied.

ahnitz commented:
@spxiwh I believe Ryan has just pushed the updated glue. Can you test that and then merge?

spxiwh commented:
Testing a short end-to-end workflow now.

@duncanmmacleod Why does `glue.version` not report a release number for the glue release?

spxiwh commented:
From testing I can see that I won't be able to push this without also pushing the associated fix for pycbc_glue and making a new release. Then PyCBC would have to depend on that pycbc_glue release. The reason for this is that the ligolw modules in pycbc_glue are currently importing `pycbc_glue.segments`, which is causing compatibility issues. Rather than hacking code to deal with this, it is easier to just do all of this in one go.

spxiwh commented:
@ahnitz @duncan-brown I think everything is ready now, as discussed on the Slack channel, I've moved the remaining part of pycbc-glue into pycbc temporarily. I've dealt with some Codacy issues and fixed a documentation build failure. (This should probably be re-approved given the changes).

However, there's still one remaining failure ... in building lalsuite for one of the OSes. I think this is some problem unrelated to this patch ... Any ideas??

ahnitz commented on pycbc/workflow/core.py:
```diff
@@ -48,7 +49,7 @@ class ContentHandler(ligolw.LIGOLWContentHandler):
 # workflow should never be using the glue LIGOTimeGPS class, override this with
 # the nice SWIG-wrapped class in lal
-lal.LIGOTimeGPS = lalswig.LIGOTimeGPS
+gluelal.LIGOTimeGPS = lal.LIGOTimeGPS
```
This should come from swig not glue. I think we wanted to move away from glue here, not back to it.

duncanmmacleod replied:
@ahnitz, in the updated code `lal` really is `lal` (and not `glue.lal`), so this line change reflects the renamed module imports.

ahnitz replied:
Ah, right. Wouldn't it make more sense to change the following lines then to directly use lal.LIGOTimeGPS then?

duncanmmacleod replied:
@ahnitz, I think the point here is that `glue.lal` has internal references to its own `LIGOTimeGPS` implementation, so you need to overwrite that if you want to have consistency amongst all things `LIGOTimeGPS`. @spxiwh can confirm.

ahnitz replied:
@duncanmmacleod  Indeed, I think you are right about that, and this line is still needed. I'm referring to the later lines, that use the LIGOTimeGPS object.

spxiwh replied:
This *should* be lal, not gluelal, as glue declares that glue.lal will soon not point to lal's LIGOTimeGPS

But glue will always be using the lal code here anyway (although it does have a fallback, that is actually broken and cannot be used). So there is no old-school glue LIGOTimeGPS with the current glue, in *any* circumstance.

spxiwh replied:
Resolved this by mostly using the lal CacheEntry and LIGOTimeGPS classes instead of ones provided by glue in this module ... I note that gluelal.Cache doesn't seem to have an equivalent in lal.utils, and this is needed to support "old-school" datafind usage (ie. when cache-files are needed).

ahnitz commented on requirements.txt:
```diff
@@ -12,7 +12,7 @@ h5py>=2.5
 mpld3>=0.3
 pyRXP>=2.1.0
-pycbc-glue-obsolete==1.1.0
+lscsoft-glue>=1.5.6
```
This needs to be updated to 1.58.2

spxiwh replied:
Yep

ahnitz commented on setup.py:
```diff
@@ -74,7 +74,7 @@
                       'mpld3>=0.3',
                       'pyRXP>=2.1.0',
-                      'pycbc-glue-obsolete==1.1.0',
+                      'lscsoft-glue>=1.56.0',
```
SHould be 
1.58.2

spxiwh replied:
Yep

ahnitz commented on setup.py:
```diff
@@ -328,7 +328,7 @@ def finalize_options(self):
         subprocess.check_call("mkdir -p _gh-pages/latest && touch _gh-pages/.nojekyll && "
                               "cd docs; cp Makefile.gh_pages Makefile; cp conf_std.py conf.py; sphinx-apidoc "
-                              " -o ./ -f -A 'PyCBC dev team' -V '0.1' ../pycbc && make html",
+                              " -o ./ -f -A 'PyCBC dev team' -V '0.1' ../pycbc ../pycbc/ligolw && make html",
```
What's the point of this? Is this building documentation for ligolw? If this is temporary, probably best not to expose this, as hopefully, ligolw will be used from normal glue or not all, no?

spxiwh replied:
This explicitly turns off documenting the ligolw module (the syntax is strange). The documentation would fail anyway as the docstrings for functions are not in sphinx formatting.

This change is needed.

ahnitz commented on setup.py:
```diff
@@ -509,9 +508,30 @@ def run(self):
+    package_data = {'pycbc.workflow': find_package_data('pycbc/workflow'),
+                    'pycbc.results': find_package_data('pycbc/results'),
+                    'pycbc.tmpltbank': find_package_data('pycbc/tmpltbank')},
+    ext_modules = [
```
Ughh.. can we not build these? I really don't like adding this in. Do they depend on lalsuite at all or are they fully self contained?

spxiwh replied:
This is self-contained C-code in glue (does not depend on lal). If we want to remove this, the best approach IMO would be to move to stage 2 of this and start figuring out why our codes are broken against the current ligolw. I don't want to do anything special here for something that is temporary. As this has been in pycbc-glue for some time, we've been doing this for a while.

spxiwh replied:
(Or self-contained C-code that's now in the ligolw module).

ahnitz commented on tools/static/hooks/hook-pycbc.py:
```diff
@@ -57,7 +57,9 @@ def find_lib_path(libname, packages):
                  'mpld3',
-                 'M2Crypto'
+                 'M2Crypto',
+                 'lal',
```
What is lal here? If it's the swig libraries, we shouldn't need this.

duncanmmacleod replied:
@ahnitz, @spxiwh, **`lscsoft-glue>=1.56` does not depend on `M2Crypto`**, and correctly declares its own dependencies on `six`, `pyOpenSSL`, and `numpy`.

spxiwh replied:
* I think lal was added here to try to fix things that are now fixed in glue. So I've reverted the change.

 * @duncanmmacleod Good point about M2Crypto, I've removed all references to this from PyCBC's build test stuff.

spxiwh commented:
Let me go through the comments .... I think some of this is rebase problems.

spxiwh commented:
@ahnitz Changes in response to comments pushed.
